src/desugar.rs

```diff
@@ -309,64 +274,112 @@ fn flatten_actions(actions: &Vec<Action>, get_fresh: &mut Fresh) -> Vec<NormActi
+    let mut res = vec![];
+    for fact in facts {
+        let mut name_used_immediately: HashSet<Symbol> = Default::default();
+        let mut constraints_before = vec![];
+        let new_fact = fact.map_def_use(&mut |var, is_def| {
+            if is_def {
+                if name_used.insert(var) {
+                    name_used_immediately.insert(var);
+                    var
+                } else {
+                    let fresh = (desugar.get_fresh)();
+                    // typechecking BS- for primitives
+                    // we need to define variables before they are used
+                    if name_used_immediately.contains(&var) {
+                        constraints.push(NormFact::ConstrainEq(fresh, var));
+                    } else {
+                        constraints_before.push(NormFact::ConstrainEq(fresh, var));
+                    }
+                    fresh
+                }
+            } else {
+                var
+            }
+        });
+        res.extend(constraints_before);
+        res.push(new_fact);
+    }
+
+    res.extend(constraints);
+    res
 }
 
-fn flatten_rule(rule_in: Rule, globals: &HashSet<Symbol>) -> NormRule {
-    let rule = parenthesize_globals(rule_in, globals);
-    let mut varcount = 0;
-    let mut get_fresh = move || {
-        varcount += 1;
-        Symbol::from(format!("fvar{}__", varcount))
-    };
+fn flatten_rule(rule: Rule, desugar: &mut Desugar) -> NormRule {
+    let flat_facts = flatten_facts(&rule.body, desugar);
+    let with_unique_names = give_unique_names(desugar, flat_facts);
 
-    let res = NormRule {
-        head: flatten_actions(&rule.head, &mut get_fresh),
-        body: flatten_facts(&rule.body, &mut get_fresh),
-    };
-    assert_ssa_valid(&res.body, &res.head);
-    res
+    NormRule {
+        head: flatten_actions(&rule.head, desugar, false),
+        body: with_unique_names,
+    }
 }
 
-pub(crate) struct Desugar {
-    pub(crate) globals: HashSet<Symbol>,
-    pub(crate) get_fresh: Box<Fresh>,
+fn desugar_run_config(desugar: &mut Desugar, run_config: &RunConfig) -> NormRunConfig {
+    let RunConfig {
+        ruleset,
+        limit,
+        until,
+    } = run_config;
+    NormRunConfig {
+        ruleset: *ruleset,
+        limit: *limit,
+        until: until.clone().map(|facts| flatten_facts(&facts, desugar)),
+    }
+}
+
+pub struct Desugar<'a> {
+    pub get_fresh: Box<Fresh>,
+    pub get_new_id: Box<NewId>,
+    pub egraph: &'a mut EGraph,
+    pub define_memo: HashMap<Expr, Symbol>,
 }
 
 pub(crate) fn desugar_command(
-    egraph: &EGraph,
     command: Command,
     desugar: &mut Desugar,
+    get_all_proofs: bool,
 ) -> Result<Vec<NormCommand>, Error> {
-    Ok(match command {
+    let res = match command {
         Command::Function(fdecl) => {
-            vec![NormCommand::Function(fdecl)]
+            vec![NCommand::Function(fdecl)]
         }
         Command::Datatype { name, variants } => desugar_datatype(name, variants),
```

Comment:
Another big change is the `declare` keyword, which declare an unbound variable.
By default, if you define a function with no arguments in `datatype`, it desugars into a `declare`.
This fixes the ambiguity between functions with no arguments and global declarations/definitions.

Previously, you could refer to functions with no arguments without parenthesis (which is evil)